### DISCUSSION TITLE
Datatable - JS-integration problems

### DISCUSSION THREAD

**Discussion by SergejGleithmann** [mdbootstrap/TW-Elements#2172] [Support from community] [answered]
Hello,
I have the following problem: 
I tried to set up a datattabel via JS as described here: https://tw-elements.com/docs/standard/data/datatables/#async-data
However JS refuses to use the constructor of the Datatable, so 
```js
const asyncTable = new te.Datatable(
        document.getElementById('datatable'),
        { columns, },
        { loading: true }
    );
```
results in `Uncaught TypeError: 'querySelector' called on an object that does not implement interface Element.`.

Can someone help me? What am I missing?

Edit: Without changing anything I sometimes also get either this: `Uncaught TypeError: te.Datatable is not a constructor` and `Uncaught TypeError: i is null`. While the first error points to my use of the Datatable constructor, the second one points to the `template.js`-File in the Datepicker-implementation of TW-Elements (yes not Datatable).

**Comment by juujisai** [comment]
Hello! I cannot reproduce those errors on my vanilla JS app with vite. Does the issue occur with the code from our Async Data example or is it your own code? https://tw-elements.com/docs/standard/data/datatables/#async-data

What stack are you using? Have you tried to use `Datatable` instead of `te.Datatable` ? Maybe your app is using the ES version of TW Elements

**Reply by SergejGleithmann** [reply]
Hello! Thanks for your answer. When I wrote the question, I had a forgotten to set the `id` of the `div` element. Now the error is different. It throws `s is undefined` consistently on first load, but if I reaload, I get `te.Datatable is not a constructor`.

I use Django and my code looks like the following:

```html
{% extends "search/base.html" %}
{% load static %}
{% load i18n %}
{% load runes_extras %}
{% block content %}
    <div class="p-2">
<!--...-->
        <div id="datatable"
             data-te-datatable-init
             data-te-class-row="py-4 pl-1 text-clip"
             data-te-class-row-item="text-clip px-[1.4rem] py-4 "
             data-te-all-text="{% translate "Alle" %}"
             data-te-rows-text="{% translate "Einträge pro Seite:" %}"
             data-te-of-text="{% translate "von" %}"
             data-te-sm="true"
             data-te-hover="true"
             data-te-full-pagination="true"></div>
    </div>
{% endblock content %}
{% block scripts %}
    <script type="module" src="/static/js/filter-runes.js"></script>
    <script>
    const lang = "{{ LANGUAGE_CODE }}";
    const filter_only = false;

    function postponeSetUp() {
        setTimeout(() => {
            if (typeof te === 'undefined') {
                postponeSetUp();
            } else {
                setUp();
            }
        }, 200);
    }

    function setUp() {
        const columns = [
            {% for bez in bez_list %}{ label: '{{ bez|translate_db }}', field: '{{ bez.key }}'},
            {% endfor %}
        ];
        
        const asyncTable = new te.Datatable(
            document.getElementById('datatable'),
            { columns, },
            { loading: true }
        );

        const queryString = window.location.search;
        window.onload = () => {
            fetch('{% url "search:api-query" %}' + queryString)
                .then((response) => response.json())
                .then((data) => {
                    asyncTable.update(
                    {
                        rows: data.finds.map((find) => ({
                        ...find,
                        {% for bez in bez_list %}
                        {{ bez.key }}: find.{{ bez.key }},
                        {% endfor %}
                        })),
                    },
                    { loading: false }
                    );
            });}
    }
    postponeSetUp();
    </script>
{% endblock scripts %}

```

The `postponeSetup`-part did not really help. I thought maybe the problem stemmed from asynchronicity. 

The TWE-JS-file is loaded in the base Template, that this file extends. It is included *before* the script section of this file as recommended in the Django set up guide for TWE.

`base.html`:
```html
<script src="{% static 'tw-elements/dist/js/tw-elements.umd.min.js' %}"
        type="text/javascript"></script>
{% block scripts %}
{% endblock scripts %}
```

**Reply by juujisai** [reply]
I think the issue may be with the timing in loading of the scripts. Check if adding `type="module"` to the style element of your `block scripts` make any difference

    <script type="module">
    const lang = "{{ LANGUAGE_CODE }}";
    ...

**Reply by juujisai** [reply]
`console.log(te.Datatable)` should log the Datatable class inside your console if everything works properly

**Reply by SergejGleithmann** [reply]
Thank you for the response. I implemented what you recommended, but it did not change anything. If I load the page (pressing enter in the URL-input) I get this error:
```
Uncaught (in promise) TypeError: s is undefined
    Ma templates.js:129
    removeDataAttribute templates.js:129
    _clearClassList popconfirm.js:239
    _clearClassList popconfirm.js:239
    update popconfirm.js:239
    fetchData table2:496
    promise callback*fetchData table2:495
    setUp table2:521
    postponeSetUp table2:466
    setTimeout handler*postponeSetUp table2:462
    EventHandlerNonNull* table2:524
templates.js:129:65
    Ma templates.js:129
    removeDataAttribute templates.js:129
    _clearClassList popconfirm.js:239
    forEach self-hosted:203
    _clearClassList popconfirm.js:239
    update popconfirm.js:239
    fetchData table2:496
    (Async: promise callback)
    fetchData table2:495
    setUp table2:521
    postponeSetUp table2:466
    (Async: setTimeout handler)
    postponeSetUp table2:462
    (Async: EventHandlerNonNull)
    <anonym> table2:524
```
In this case `console.log(te.Datatable)` works.
And if I reload the page, the error is 
```
Uncaught TypeError: te.Datatable is not a constructor
    setUp http://localhost:8000/table2?table=onfund&filter=runenreihe_de eq älteres fuþark&select=findnr,hauptsigle,translit,runenreihe_de,fundort_de,land,landschaft&order=findnr asc&skip=80&top=20:484
    postponeSetUp http://localhost:8000/table2?table=onfund&filter=runenreihe_de eq älteres fuþark&select=findnr,hauptsigle,translit,runenreihe_de,fundort_de,land,landschaft&order=findnr asc&skip=80&top=20:466
    setTimeout handler*postponeSetUp http://localhost:8000/table2?table=onfund&filter=runenreihe_de eq älteres fuþark&select=findnr,hauptsigle,translit,runenreihe_de,fundort_de,land,landschaft&order=findnr asc&skip=80&top=20:462
    EventHandlerNonNull* http://localhost:8000/table2?table=onfund&filter=runenreihe_de eq älteres fuþark&select=findnr,hauptsigle,translit,runenreihe_de,fundort_de,land,landschaft&order=findnr asc&skip=80&top=20:524
```

**Reply by juujisai** [reply]
What if we change the type here from `type="text/javascript"` to `type="module"` ? I still think the issue is with the time the files load. We had similar issue with our documentation, adding those types helped in our case.

```
<script src="{% static 'tw-elements/dist/js/tw-elements.umd.min.js' %}"
        type="module"></script>
```

**Reply by SergejGleithmann** [reply]
@juujisai Thanks for the reply, sorry for my late response. Sadly changing the type does not change the behaviour, but I also think, that it has to do with the timing, because `te` exists if the page is loaded.
![TE](https://github.com/mdbootstrap/TW-Elements/assets/110982970/7def9fd9-390e-49c8-82d2-f8ecd86df20a)

**Reply by iprzybysz** [reply]
Hi @SergejGleithmann, 

firstly try changing the time of loading your scripts to `module`.
```
<script type="module">
const lang = "{{ LANGUAGE_CODE }}";
...
```

If this doesn't help, you can try out the [DOMContentLoaded](https://developer.mozilla.org/en-US/docs/Web/API/Document/DOMContentLoaded_event) event and when a document is fully loaded, initialise the Datatable component. You can also check the import order in the `base.html` file. I recommend you check our [Django Integration tutorial](https://tw-elements.com/docs/standard/integrations/django-integration/) and make sure you've done everything right.

I've created a new Django App based on the [Django Integration tutorial](https://tw-elements.com/docs/standard/integrations/django-integration/) and everything works for me. Here is my code - maybe it'll also help you with debugging your code:

**base.html**
```
{% load static %}
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link href="{% static 'output.css' %}" rel="stylesheet" />
    <title>
      {% block head_title %}TWE Django Integration Tutorial{% endblock %}
    </title>
  </head>
  <body>
    <div>{% block content %} {% endblock %}</div>
    <script
      src="{% static 'tw-elements/dist/js/tw-elements.umd.min.js' %}"
      type="text/javascript"
    ></script>
  </body>
</html>
```

**home.html**
```
{% extends "base.html" %} {% load static %} {% block content %}
<div id="datatable"></div>

<script type="module">
  const basicData = {
    columns: ['Name', 'Position', 'Office', 'Age', 'Start date', 'Salary'],
    rows: [
      ["Tiger Nixon", "System Architect", "Edinburgh", "61", "2011/04/25", "$320,800"],
      ["Garrett Winters", "Accountant", "Tokyo", "63", "2011/07/25", "$170,750"],
      ["Ashton Cox", "Junior Technical Author", "San Francisco", "66", "2009/01/12", "$86,000"],
      ["Cedric Kelly", "Senior Javascript Developer", "Edinburgh", "22", "2012/03/29", "$433,060"],
      ["Airi Satou", "Accountant", "Tokyo", "33", "2008/11/28", "$162,700"],
      ["Brielle Williamson", "Integration Specialist", "New York", "61", "2012/12/02", "$372,000"],
      ["Herrod Chandler", "Sales Assistant", "San Francisco", "59", "2012/08/06", "$137,500"],
      ["Rhona Davidson", "Integration Specialist", "Tokyo", "55", "2010/10/14", "$327,900"],
      ["Colleen Hurst", "Javascript Developer", "San Francisco", "39", "2009/09/15", "$205,500"],
      ["Sonya Frost", "Software Engineer", "Edinburgh", "23", "2008/12/13", "$103,600"],
      ["Jena Gaines", "Office Manager", "London", "30", "2008/12/19", "$90,560"],
      ["Quinn Flynn", "Support Lead", "Edinburgh", "22", "2013/03/03", "$342,000"],
      ["Charde Marshall", "Regional Director", "San Francisco", "36", "2008/10/16", "$470,600"],
      ["Haley Kennedy", "Senior Marketing Designer", "London", "43", "2012/12/18", "$313,500"]
    ],
  };

  new te.Datatable(document.getElementById("datatable"), basicData);
</script>
{% endblock %}
```

**Reply by SergejGleithmann** [reply, marked as answer]
tl;dr: The `tw-elements.umd.min.js` script-tag needs to be of type `text/javascript`, but additional scripts need to be of type `module`. This way `te` (and all the TWE features) is available, when the additional script runs.

Hi @iprzybysz,
thank you for your response. It put me on the right track. 

As @juujisai proposed earlier the `type`-attribute was the problem. However, in my case I needed my js code (the code in the `home.html` file) to be of type `module` - not the Tailwind-JS file.

Thank you all for your help!
